proposals/0000-warning-pragmas-with-categories.rst

```diff
@@ -0,0 +1,332 @@
+
+A warning *category* is an identifier beginning with a letter and followed by
+valid identifier characters or dashes.
+
+#. A `WARNING pragma
+   <https://downloads.haskell.org/ghc/9.4.1/docs/users_guide/exts/pragmas.html#warning-deprecated-pragma>`_
+   may be immediately followed by a single warning category in square brackets.
+   See below for details of the grammatical changes.
+
+#. Individual warning categories may be enabled or disabled using new
+   ``-W<category>`` or ``-Wno-<category>`` options, and their priority may be
+   controlled using the ``-Werror=<category>`` or ``-Wwarn=<category>`` options.
+
+#. A ``WARNING`` pragma without a category, or a ``DEPRECATED`` pragma, is
+   interpreted as if it was a ``WARNING`` pragma with the single category
+   ``deprecations`` specified.
+
+#. The existing ``-Wwarnings-deprecations`` warning flag is interpreted as a
+   synonym for ``-Wdeprecations`` (and similarly for
+   ``-Wno-warnings-deprecations``, ``-Werror=warnings-deprecations`` and so on).
+
+#. A new warning flag ``-Wextended-warnings`` switches on warnings from
+   ``WARNING`` pragmas regardless of category; ``-Wno-extended-warnings``
+   switches them all off.
+
+#. A category is *recognised* if it either starts with ``x-`` or is
+   ``deprecations``.  GHC will emit an error if a ``WARNING`` pragma uses a
+   category that is not recognised.
+
+There is no change to the existing rules for when ``WARNING`` pragmas give rise
+to warnings; these changes merely affect whether the warnings are displayed.
+The command-line flags are processed from left to right, with later
+flags overriding previous ones, as at present.
+
+Points 3 and 4 make the proposal backwards compatible: existing ``WARNING`` or
+``DEPRECATED`` pragmas will still be controlled via the ``-Wdeprecations`` or
+``-Wwarnings-deprecations`` options.  However if a library introduces a category
+for a previously uncategorised warning, the warning will no longer be suppressed
+by ``-Wno-deprecations``.  (Prior to this proposal, all ``WARNING`` and
+``DEPRECATED`` pragmas were treated uniformly and could be disabled either by
+``-Wno-deprecations`` or by ``-Wno-warnings-deprecations``.)
+
+Point 5 is not strictly necessary, but ``-Wno-extended-warnings`` will allow
+users to suppress all ``WARNING`` messages much like
+``-Wno-warnings-deprecations`` does at present.  The naming is chosen to allow
+for other sources of "extended warnings" in the future.
+
+The restriction to recognised categories in point 6 means that ``WARNING``
+pragmas cannot extend the behaviour of existing GHC warning flags such as
+``-Wtabs`` or groups such as ``-Wall``.  Moreover it means that GHC can still
+report unrecognised warning flags, rather than silently accepting them.  In the
+future we may extend the set of recognised categories, if it appears useful to
+let users extend specific built-in warning categories (e.g. ``-Wcompat``) or to
+allow prefixes other than ``x-``.
```

Comment:
```suggestion
Point 6 makes sure that the command line options `-Wx-partial` and `-Wno-x-partial`
can readily be distinguished from the existing large number of built-in warning
categories, such as `-Wtabs`, `-Wdodgy-import`, `-Winaccessible-code` etc. (See the
`user manual section <https://ghc.gitlab.haskell.org/ghc/doc/users_guide/using-warnings.html?highlight=warning#warnings-and-sanity-checking>`_ for a complete list).
This way GHC can still
report unrecognised warning flags, rather than silently accepting them.
```